### PR TITLE
comment activities

### PR DESCRIPTION
**Pull request by wilmerterrero**
## Description
Creating functionality to add comments to activities. But the comment form is not showing.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] I verified that my code applies to the guidelines (`composer code-check`)
- [ ] I updated the documentation (see [here](https://github.com/kimai/www.kimai.org/tree/master/_documentation))
- [x] I agree that this code is used in Kimai and will be published under the [MIT license](https://github.com/kevinpapst/kimai2/blob/master/LICENSE)

### CODE REVIEW

**Comment by wilmerterrero** [comment]
I created the table for the comments of the activities.
```sql
-- phpMyAdmin SQL Dump
-- version 5.0.2
-- https://www.phpmyadmin.net/
--
-- Host: 127.0.0.1
-- Generation Time: Sep 22, 2020 at 05:35 PM
-- Server version: 10.4.14-MariaDB
-- PHP Version: 7.4.10

SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
START TRANSACTION;
SET time_zone = "+00:00";


/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8mb4 */;

--
-- Database: `kimai_beneli`
--

-- --------------------------------------------------------

--
-- Table structure for table `kimai2_activities_comments`
--

CREATE TABLE `kimai2_activities_comments` (
  `id` int(11) NOT NULL,
  `activity_id` int(11) NOT NULL,
  `created_by_id` int(11) NOT NULL,
  `message` longtext COLLATE utf8mb4_unicode_ci NOT NULL,
  `created_at` datetime NOT NULL COMMENT '(DC2Type:datetime)',
  `pinned` tinyint(1) NOT NULL DEFAULT 0
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;


--
-- Indexes for table `kimai2_activities_comments`
--
ALTER TABLE `kimai2_activities_comments`
  ADD PRIMARY KEY (`id`),
  ADD KEY `IDX_ACTIVITY_COMMENT_ID` (`activity_id`),
  ADD KEY `IDX_ACTIVITY_COMMENT_BY_ID` (`created_by_id`);

--
-- AUTO_INCREMENT for dumped tables
--

--
-- AUTO_INCREMENT for table `kimai2_activities_comments`
--
ALTER TABLE `kimai2_activities_comments`
  MODIFY `id` int(11) NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=1;

--
-- Constraints for dumped tables
--

--
-- Constraints for table `kimai2_activities_comments`
--
ALTER TABLE `kimai2_activities_comments`
  ADD CONSTRAINT `FK_ACTIVITY_COMMENT_ID` FOREIGN KEY (`activity_id`) REFERENCES `kimai2_activities` (`id`) ON DELETE CASCADE,
  ADD CONSTRAINT `FK_ACTIVITY_COMMENT_USER_ID` FOREIGN KEY (`created_by_id`) REFERENCES `kimai2_users` (`id`) ON DELETE CASCADE;
COMMIT;

/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
/*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
/*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;
```

**Comment by kevinpapst** [comment]
I guess you have seen the [found issues](https://github.com/kevinpapst/kimai2/pull/1986/files) from the build. Most important: the repository is missing the required methods ... once you have updated the PR, you can ping me and I will have a look.

**Comment by wilmerterrero** [comment]
@kevinpapst Can you check to see what happens?
I guess everything is fine. But the form is still not showing.

File: src/Controller/ActivityController.php [hunk 1]
```diff
@@ -195,6 +209,61 @@ public function createAction(Request $request, ?Project $project = null)
         return $this->renderActivityForm($activity, $request);
     }
 
+    /**
+     * @Route(path="/{id}/comment_delete", name="project_comment_delete", methods={"GET"})
```
Comment (kevinpapst) [inline src/Controller/ActivityController.php:213]: wrong route name

File: src/Controller/ActivityController.php [hunk 2]
```diff
@@ -195,6 +209,61 @@ public function createAction(Request $request, ?Project $project = null)
...
+            } catch (\Exception $ex) {
+                $this->flashError('action.update.error', ['%reason%' => $ex->getMessage()]);
+            }
+        }
+
+        return $this->redirectToRoute('activity_details', ['id' => $activity->getId()]);
+    }
+
+    /**
+     * @Route(path="/{id}/comment_pin", name="project_comment_pin", methods={"GET"})
```
Comment (kevinpapst) [inline src/Controller/ActivityController.php:252]: wrong route name

File: src/Entity/ActivityComment.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+namespace App\Entity;
+
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\Validator\Constraints as Assert;
+
+/**
+ * @ORM\Entity()
+ * @ORM\Table(name="kimai2_activity_comments",
```
Comment (kevinpapst) [inline src/Entity/ActivityComment.php:19]: wrong table name, must be `kimai2_activities_comments`

File: templates/activity/details.html.twig [hunk 4]
```diff
@@ -93,6 +93,14 @@
         {{ include('embeds/teams.html.twig', options) }}
     {% endif %}
 
+    {% if comments is not null %}
+        {% set options = {'form': commentForm, 'comments': comments} %}
+        {% if can_edit %}
+            {% set options = options|merge({'route_pin': 'project_comment_pin', 'route_delete': 'project_comment_delete'}) %}
```
Comment (kevinpapst) [inline templates/activity/details.html.twig:99]: wrong route names

**Review by kevinpapst — CHANGES_REQUESTED** [review]
Okay, next step: you need to adjust the [Activity voter](https://github.com/kevinpapst/kimai2/blob/master/src/Voter/ActivityVoter.php) like this:
https://github.com/kevinpapst/kimai2/blob/master/src/Voter/ProjectVoter.php#L31-L32
otherwise no one has permissions to see the comments.

File: templates/activity/details.html.twig (hunk 4 above)
Comment (wilmerterrero) [reply templates/activity/details.html.twig:99]: Now works!!! Thanks man
image

**Comment by kevinpapst** [comment]
Do you plan to improve the PR and make it production ready for Kimai?
I could guide you in that case. Otherwise the PR will stay here until I find the time to work on it in the future.

**Comment by wilmerterrero** [comment]
> Do you plan to improve the PR and make it production ready for Kimai?
> I could guide you in that case. Otherwise the PR will stay here until I find the time to work on it in the future.

Yes I can do it. The only problem is I dont know why is falling the test.
Could you guide me with that?

**Comment by kevinpapst** [comment]
Thanks for the contribution so far!
Due to housekeeping I am going to close this issue. If you or someone else wants to further work on it, please go ahead and rebase on master + open a new PR.

**Pull request closed without merging**